### PR TITLE
Test lakefsfs with mock server

### LINKED ISSUE #6554: Refactor LakeFSFileSystemTest to use a mock server rather than mockito
Modify LakeFSFileSystemTest to test that lakeFSFS works with an appropriate REST API.

The current LakeFSFileSystemTest mocks the existing API client.  For #6424 we will be switching the generated API.  This means we will need to rewrite _all_ mocking to match the new API.

**It gets worse:** the new API has a fluent style, and Mockito can mock fluent styles but only in a non-composable style.  So we will anyway need to rewrite the most complex parts of the system test.

### PR DESCRIPTION
**Pull request by arielshaqed**
Translate almost all tests in LakeFSFileSystemTest to use a Java MockClient.  The result is called LakeFSFileSystemServerTest.

The existing Mockito-based code is horribly brittle; see the underlying issue #6554 for  why this manages to _block_ our move to a forwards-compatible Java SDK.  So this is blocking #6424 .

Fixes #6554 _for the most part_.  Some tests are still missing, but I do not want to block the Java client review on fixing them.  I will open a new issue to uncomment these tests, and copy over any tests from LakeFSFileSystemTest that are still missing.

### CODE REVIEW

File: clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemServerTest.java [hunk 1]
```diff
@@ -0,0 +1,1375 @@
...
+    }
+
+    @Test
+    public void getUri() {
+        URI u = fs.getUri();
+        Assert.assertNotNull(u);
+    }
+
+    // Expect this statObject to be not found
+    protected void expectStatObjectNotFound(String repo, String ref, String path) {
```
Comment (johnnyaug) [inline clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemServerTest.java:293]: Not really an expectation: it's setting the mock to return NotFound. Same for the next ones.

File: clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemServerTest.java [hunk 2]
```diff
@@ -0,0 +1,1375 @@
...
+        expectStatObjectNotFound("repo", "main", "non-existing/");
+
+        boolean renamed = fs.rename(src, dst);
+        Assert.assertFalse(renamed);
+    }
+
+    @Test
+    public void testRename_existingFileToExistingFileName() throws IOException {
+        String contents = "The quick brown fox jumps over the lazy dog.";
+        byte[] contentsBytes = contents.getBytes();
```
Comment (johnnyaug) [inline clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemServerTest.java:1099]: Used?

File: clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemServerTest.java [hunk 3]
```diff
@@ -0,0 +1,1375 @@
...
+                     .withStatusCode(201)
+                     // Actual new dstStats will be different... but lakeFSFS doesn't care.
+                     .withBody(json(gson.toJson(dstStats))));
+        expectGetBranch("repo", "main");
+        expectDeleteObject("repo", "main", "existing-dir1/existing.src");
+
+        // Need a directory marker at the source because it's now empty!
+        expectUploadObject("repo", "main", "existing-dir1/");
+
+        Assert.assertTrue(fs.rename(src, dst));
```
Comment (johnnyaug) [inline clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemServerTest.java:1193]: Many tests here (and their original counterparts) do not test much. Can we test the physical file at this point, similar to what we do on testCreate, for example?

**Review by johnnyaug — CHANGES_REQUESTED** [review]
Nothing major.

**Review by arielshaqed — COMMENTED** [review]
I refactored S3 tests in [f8372ad](https://github.com/treeverse/lakeFS/pull/6646/commits/f8372ad4614ef7679c13e2fce31ee747115ca8b0), then used that in [e0b075d](https://github.com/treeverse/lakeFS/pull/6646/commits/e0b075df9c3067d316521124e1b597345ebb379d) to test presigned and simple modes (they work!).

Other review comments are on lines that I've already clobbered, so responding on them _there_; sorry!

File: clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemTest.java [hunk 4]
```diff
@@ -925,7 +925,7 @@ public void testRename_existingFileToExistingDirName() throws ApiException, IOEx
         ObjectLocation fileObjLoc = fs.pathToObjectLocation(fileInDstDir);
         Path dst = new Path("lakefs://repo/main/existing-dir2");
         ObjectLocation dstObjLoc = fs.pathToObjectLocation(dst);
-        mockExistingDirPath(dstObjLoc, ImmutableList.of(fileObjLoc));
+        //        mockExistingDirPath(dstObjLoc, ImmutableList.of(fileObjLoc));
```
[the remark this answers is not in the document]
Comment (arielshaqed) [reply clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemTest.java]: Apparently it was never needed(!), the code discovers that the file exists even before listing it 🤷🏽 .

File: clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemServerTest.java (hunk 1 above)
Comment (arielshaqed) [reply clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemServerTest.java]: I hate the word "mock", but changing words back.

File: clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemServerTest.java (hunk 2 above)
Comment (arielshaqed) [reply clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemServerTest.java:516]: Yikes.  It's not!  It wasn't before, but it still isn't used.

File: clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemServerTest.java (hunk 3 above)
Comment (arielshaqed) [reply clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemServerTest.java:610]: There are issues with this test, agreed.  As they do not block 1.0, I do not want to perform them here.

Note however that using a mockserver tightens up these tests!  The "smart nulls" behaviour that we used on Mockito means that many calls will fail silently and _differently_ to how they fail on a real HTTP server.

File: clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (johnnyaug) [reply clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemTest.java]: Let's delete it then!

File: clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemServerS3Test.java [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+
+    @Override
+    protected void moreHadoopSetup() {
+        super.moreHadoopSetup();
+        pac.initConfiguration(conf);
+    }
+
+    // Return a location under namespace for this getPhysicalAddress call.
+    //
+    // TODO(ariels): abstract, overload separately for direct and pre-signed.
```
Comment (johnnyaug) [inline clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemServerS3Test.java:116]: TODO is already done?

File: clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemServerS3Test.java [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+import static org.mockserver.model.HttpResponse.response;
+import static org.mockserver.model.JsonBody.json;
+
+import java.io.*;
+import java.net.URL;
+import java.util.Date;
+import java.util.Arrays;
+import java.util.concurrent.TimeUnit;
+
+@RunWith(Parameterized.class)
```
Comment (johnnyaug) [inline clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemServerS3Test.java:37]: Neat!

**Comment by johnnyaug** [comment]
Tests that are no longer there:
```
testRename_existingDirToExistingNonEmptyDirName
testRename_fallbackStageAPI
testRename_nonExistingSrcFile
testRename_srcAndDstOnDifferentBranch
testRename_srcEqualsDst
```

Are we cool with that?

**Review by johnnyaug — APPROVED** [review]
See some minor comments

**Review by arielshaqed — COMMENTED** [review]
Thanks!  I think no major major changes in this, and you approved, so I will pull once tests pass.  This _does_ of course mean that you get to ask for more changes if I did something horribly wrong 🤷🏽 .

**Comment by arielshaqed** [comment]
# THANKS!

## Sorry for the mega-review

If it's any consolation, #6656 is even worse!

Pulling once tests pass.

**Pull request merged**